gotest_me/BBQ_test.go

```diff
@@ -0,0 +1,116 @@
+
+	t.Run("calc", func(t *testing.T) {
+		t.Run("when people counter is exceeded max value", func(t *testing.T) {
+			defer initMocks(t)()
+			_, err := subject.CalculatePrice(11)
+			assert.NotNil(t, err)
+		})
+		t.Run("when people counter is not exceeded max value",func(t *testing.T) {
+			defer initMocks(t)()
+			t.Run("when get entrecote returns error", func(t *testing.T) {
+				peopleCounter := 10
+				mockBeefFarm.EXPECT().GetEntrecote(peopleCounter).Return(0, errors.New("test_err_7483")).Times(1)
+
+				t.Run("when get chicken returns error", func(t *testing.T) {
+					mockChickenFarm.EXPECT().GetPullet(peopleCounter).Return(0, errors.New("test_err_2393"))
+					_, err := subject.CalculatePrice(peopleCounter)
+					assert.NotNil(t, err)
+				})
+				t.Run("when get chicken returns no error", func(t *testing.T) {
+					t.Run("when mangal return error", func(t *testing.T) {
+						// In gincko test this text
+						// this case is not tested
+						// flow is the same as next "when get entrecote returns no error"
+						// but I think we need test this case too. Because we must be sure that meetPrice was overwritten
+					})
+				})
+			})
+
+			t.Run("when get entrecote returns no error", func(t *testing.T) {
+				defer initMocks(t)()
+				peopleCounter := 10
+				meetPrice := 100
+				mockBeefFarm.EXPECT().GetEntrecote(peopleCounter).Return(meetPrice, nil).AnyTimes()
+				t.Run("when get mangal returns error", func(t *testing.T) {
+					mockMangalStore.EXPECT().GetMangal().Return(0, errors.New("test_err_9494")).Times(1)
+
+					t.Run("should return error about no mangal available", func(t *testing.T) {
+						_, err := subject.CalculatePrice(peopleCounter)
+						assert.NotNil(t, err, )
+						assert.Contains(t, err.Error(), "no mangal available")
+						assert.Contains(t, err.Error(), "test_err_9494")
+					})
+				})
+
+				t.Run("when get mangal returns no error", func(t *testing.T) {
+					mangalPrice := 1000
+					mockMangalStore.EXPECT().GetMangal().Return(mangalPrice, nil).Times(2)
+
+					t.Run("when get coal returns error", func(t *testing.T) {
+						mockCoatFarm.EXPECT().GetCoal(1).Return(0, errors.New("test_err_9883"))
+
+						t.Run("should return error about no coal available", func(t *testing.T) {
+							_, err := subject.CalculatePrice(peopleCounter)
+							assert.NotNil(t, err, )
+							assert.Contains(t, err.Error(), "no coal available")
+							assert.Contains(t, err.Error(), "test_err_9883")
+						})
+					})
+
+					t.Run("when get coal returns no error", func(t *testing.T) {
+						coatPrice := 10
+						mockCoatFarm.EXPECT().GetCoal(1).Return(coatPrice, nil)
+
+						t.Run("should return no error and valid total price", func(t *testing.T) {
+							got, err := subject.CalculatePrice(peopleCounter)
+							assert.Nil(t, err)
+							expectedTotalPrice := meetPrice + mangalPrice + coatPrice
+							assert.Equal(t, expectedTotalPrice, got)
+						})
+					})
+				})
+			})
+		})
+	})
+}
```

Comment:
Totally looks nice - pretty same as Ginkgo. At least same structure.
I don't see need to use ginkgo if the structure reminds clear.
Important to note - here there is no clearly defined structure in defining test cases and tests sserts. 

This might make bit harder to "keep right". For example, this (last 2 `t.Run(...)`):
 ```
...
t.Run("when get coal returns no error", func(t *testing.T) {
        coatPrice := 10
        mockCoatFarm.EXPECT().GetCoal(1).Return(coatPrice, nil)

        t.Run("should return no error and valid total price", func(t *testing.T) {
                got, err := subject.CalculatePrice(peopleCounter)
                assert.Nil(t, err)
                expectedTotalPrice := meetPrice + mangalPrice + coatPrice
                assert.Equal(t, expectedTotalPrice, got)
        })
})
```
Might be written as:
```
...
t.Run("when get coal returns no error", func(t *testing.T) {
        coatPrice := 10
        mockCoatFarm.EXPECT().GetCoal(1).Return(coatPrice, nil)
			
        // removed internal t.Run...
        got, err := subject.CalculatePrice(peopleCounter)
        assert.Nil(t, err)
        expectedTotalPrice := meetPrice + mangalPrice + coatPrice
        assert.Equal(t, expectedTotalPrice, got)
})
```
I don't see clear reason why this is not good (despite it is :)
WDYT?

Reply:
i think both structure make sense as long as you dont define global variables with one value and change them during test like:  coatPrice := 10 .
the flat and nested case can be similar as long as there is no much nesting alot of nesting hides the tests flow from the reader since variables being define and used not in same place (example the just before each in ginkgo) . and since the flow initialisation order is not clear and intuitive (my opinion :) )